pkg/controller/common.go

```diff
@@ -277,14 +284,30 @@ func getRankedNodesForReplicaAttachments(ctx context.Context, azr azReconciler,
 			}
 		}
 	}
+
+	for nodeName, nodeScore := range nodeScores {
+		// it is safe to delete entry from map while iterating
+		removeCandidacy := func(nodeName string, nodeMap map[string]int, err error) {
+			klog.Errorf("Removing node (%s) from candidate nodes for replica attachments: %v", nodeName, err)
+			delete(nodeMap, nodeName)
+		}
+		if remainingCapacity, err := getNodeRemainingCapacity(ctx, azr.getClient(), nil, nodeName); err != nil {
+			removeCandidacy(nodeName, nodeScores, err)
+		} else if remainingCapacity < len(volumes)-nodeScore {
+			// if node is too full to take n extra attachments, remove the node from the candidate list
+			err = status.Errorf(codes.Internal, "remaining node capacity: %d", remainingCapacity)
+			removeCandidacy(nodeName, nodeScores, err)
```

Comment:
If we are removing a higher scoring node from consideration, we should ideally also move replicas to node that can hold them all.

Reply:
Migrating replica attachments of one pod to guarantee a fast failover for that pod might end up in slow failover for a different pod.

Say we have pod A and pod B, each with volumes set {1, 2} and {2, 3} respectively. Assume pod A came first and we successfully were able to place volume 1 and 2's replica attachments on the node a. Then pod B comes next and finds that node a is full and decides to migrate volume 2's replica to node b with sufficient capacity. By doing so, we may be able to guarantee fast failover for pod B but we lose fast failover for pod A.

One might say that then we can migrate all volume 1, 2, 3 to node b but as this inevitably involves searching pod topology, coming up with the full list of replica attachments to migrate may not be a realistic goal. And this can very easily involve a very large migration, which could a bigger loss than a gain.